### PR TITLE
feat: Add `checkJSDoc` option to multiline-comment-style

### PR DESCRIPTION
**Pull request by laurent22**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))

**What rule do you want to change?**

multiline-comment-style

**What change do you want to make (place an "X" next to just one item)?**

[ ] Generate more warnings
[ ] Generate fewer warnings
[x] Implement autofix
[x] Implement suggestions

**How will the change be implemented (place an "X" next to just one item)?**

[x] A new option
[ ] A new default behavior
[ ] Other

**Please provide some example code that this change will affect:**

With this rule: `'multiline-comment-style': ['error', 'separate-lines', { allowJsDoc: false }]`, this is not allowed:

```js
/**
 * Some
 * comment
 */
```

And is converted to this:

```js
// Some
// comment
```

By default `allowJsDoc` is `true` so all other behaviours remain the same.

The rational for adding this is that we disallow block comments in our codebase but some contributors (voluntarily or not) bypass this by formatting the block comment as a jsdoc comment. We don't want this, so this option would help with this.

I have just realised I forgot to update the documentation but happy to do so if you would consider this change.

### CODE REVIEW

**Comment by nzakas** [comment]
Thanks for this. Can you explain a bit more about the change you made here? What was the rule not doing?

**Comment by laurent22** [comment]
Sure, the issue is with this `separate-lines` option - normally it disallows block comments in favor of consecutive line comments. So this is ok:

```
// My
// comment
```

This is not:

```
/* My
   comment */
```

The problem, is that there's currently a hard-coded exception for jsdoc. So with the existing rule, this type of block comment is ok:

```
/**
 * My
 * comment
 */
```

Basically, formatting the comment as jsdoc allows bypassing the `separate-lines` option. This might be fine for some codebases but not when jsdoc is not in use (like in our codebase).

So this is why I'm proposing this `allowJsDoc` option. By default it is `true`, which means existing configurations won't be affected, but those who don't want this option can set it to `false` and disallow jsdoc-style comments too.

**Comment by aladdin-add** [comment]
Thanks for the input. As [stated in the README](https://github.com/eslint/eslint#stylistic-rule-updates), we actually aren't making any further updates to rules that enforce stylistic preferences. If you have a strong preference for such a change, you can always copy the existing rule into a custom rule and make it behave the way you want.

**Comment by laurent22** [comment]
Changing the doc wouldn't help since the `separate-lines` parameter would still not be doing what it's supposed to be dong.

**Comment by nzakas** [comment]
@laurent22 thanks for the explanation. So it sounds like what you want to do is just opt-out of the hardcoded JSDoc exception?

**Comment by nzakas** [comment]
**TSC Summary:** This PR seeks to add a way to opt-out of the hardcoded JSDoc comment exception in `multiline-comment-style` ([full description](https://github.com/eslint/eslint/pull/16807#issuecomment-1417706498)). The `multiline-comment-style` rule is stylistic and frozen, meaning we won't make any further changes to the rule, but this pull request is fully implemented.

**TSC Questions:**

1. Should we accept this pull request?
2. Should the option be named `allowJsDoc` or `allowJSDoc`?

**Comment by btmills** [comment]
In the [2023-02-23 TSC meeting](https://github.com/eslint/tsc-meetings/blob/b2f5017aeaa8ca1d40de275d936a02e8623a5d49/notes/2023/2023-02-23.md#feat-add-support-for-allowjsdoc-param-for-multiline-comment-style), we decided to accept this pull request because this rule was designed with this hard-coded exception but no way to opt out.

@mdjermanovic while preparing the notes, I noticed we never answered part 2: should it be `allowJsDoc` or `allowJSDoc`? `rg --stats JSDoc lib` reports 59 matches across seven JS files for comments and code containing `JSDoc`, whereas `rg --stats JsDoc lib` only reports five occurrences of `JsDoc`, all referencing [one `checkJsDoc` function in `lib/rules/require-jsdoc.js`](https://github.com/eslint/eslint/blob/8e34a04e3a4395bce59bc6acadf84281abc11d18/lib/rules/require-jsdoc.js#L88). Given that precedent, I think the option should be named `allowJSDoc`.

**Comment by mdjermanovic** [comment]
> while preparing the notes, I noticed we never answered part 2: should it be `allowJsDoc` or `allowJSDoc`? `rg --stats JSDoc lib` reports 59 matches across seven JS files for comments and code containing `JSDoc`, whereas `rg --stats JsDoc lib` only reports five occurrences of `JsDoc`, all referencing [one `checkJsDoc` function in `lib/rules/require-jsdoc.js`](https://github.com/eslint/eslint/blob/8e34a04e3a4395bce59bc6acadf84281abc11d18/lib/rules/require-jsdoc.js#L88). Given that precedent, I think the option should be named `allowJSDoc`.

Indeed, I missed the second part. Agreed, the option should be named `allowJSDoc`.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]
@laurent22 this PR is now accepted, but the option name should be changed to `allowJSDoc`. You can also update the documentation for this rule now.

File: lib/rules/multiline-comment-style.js
Comment (mdjermanovic) [inline lib/rules/multiline-comment-style.js:39]: Does the new option apply only when the string option is `"separate-lines"`? If so, to avoid confusion we should enforce in the schema that the option can be specified only with `"separate-lines"`.

**Comment by laurent22** [comment]
Thank you for considering the PR, I will implement the required changes as soon as possible

**Comment by laurent22** [comment]
I've now renamed the option to `allowJSDoc` and added the documentation.

File: lib/rules/multiline-comment-style.js
Comment (laurent22) [reply lib/rules/multiline-comment-style.js]: Yes it only applies to `separate-lines`. I've looked at existing rules and couldn't find an example of enforcing an option associated with another option. Do you have any example or reference I could look at to implement this?

Comment (laurent22) [reply lib/rules/multiline-comment-style.js]: Ok now that I see your commit I get it. Thanks for that

**Comment by laurent22** [comment]
I believe the pull request is ready now, and just needs to wait for CI

File: docs/src/rules/multiline-comment-style.md
Comment (nzakas) [inline docs/src/rules/multiline-comment-style.md:20]: Wording here is a little confusing, maybe this helps?
```suggestion
* `"separate-lines"`: Disallows block comments in favor of consecutive line comments. By default, this option ignores JSDoc comments. To also apply this rule to JSDoc comments, set the `allowJSDoc` option to `false`.
```

File: lib/rules/multiline-comment-style.js
Comment (nzakas) [inline lib/rules/multiline-comment-style.js:71]: It seems strange to have an option that is `true` by default.

Could we change this option to be called `ignoreJSDoc` so that the default is `false` and people can override to `true`?

@mdjermanovic what do you think?

**Review by nzakas — CHANGES_REQUESTED** [review]

File: lib/rules/multiline-comment-style.js
Comment (mdjermanovic) [reply lib/rules/multiline-comment-style.js:71]: Wouldn't the default for `ignoreJSDoc` be `true` as well?

If we want the default to be `false`, then perhaps `enforceForJSDoc` or `disallowJSDoc`? The default `disallowJSDoc: false` is a kind of double negation, so it might be confusing. But I think it's a good name since the wording allow/disallow seems suitable for this option because after converting to line comments it's no longer a valid JSDoc, so this option effectively does disallow multiline JSDoc comments.

Comment (laurent22) [reply lib/rules/multiline-comment-style.js:71]: I used `allowJSDoc`  because I prefer positive booleans in general, indeed to avoid thinking in terms of double negatives. If you prefer to change it though, please confirm which one you would prefer as I see multiple possibilities above.

Comment (nzakas) [reply lib/rules/multiline-comment-style.js:71]: Positive booleans are definitely preferable, however, not as a default value for an optional setting. Because undefined is falsy in JavaScript, its more idiomatic to have optional settings default to false.

Maybe `checkJSDoc` would work? So `checkJSDoc: false` is the default, so it leaves JSDoc comments alone, and `checkJSDoc: true` can override that and include JSDoc comments in the check?

Comment (mdjermanovic) [reply lib/rules/multiline-comment-style.js:71]: `checkJSDoc` sounds good to me.

**Comment by laurent22** [comment]
Thank you for your feedback, I have now renamed the option to `checkJSDoc` and also updated the documentation as suggested by @nzakas

File: lib/rules/multiline-comment-style.js
Comment (mdjermanovic) [inline lib/rules/multiline-comment-style.js:39]:
```suggestion
        schema: {
            anyOf: [
                {
                    type: "array",
                    items: [
                        {
                            enum: ["starred-block", "bare-block"]
                        }
                    ],
                    additionalItems: false
                },
                {
                    type: "array",
                    items: [
                        {
                            enum: ["separate-lines"]
                        },
                        {
                            type: "object",
                            properties: {
                                checkJSDoc: {
                                    type: "boolean"
                                }
                            },
                            additionalProperties: false
                        }
                    ],
                    additionalItems: false
                }
            ]
        },
```

Looks like this change was lost while rebasing.

**Comment by laurent22** [comment]
Thank you for the change suggestions @mdjermanovic. I've integrated them and ran the unit tests again, so I think it is ready now

**Review by nzakas — APPROVED** [review]
LGTM. Thanks for your patience on this. Just want @mdjermanovic to approve before merging.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Comment by nzakas** [comment]
@laurent22 we'd like to pay you for this contribution. Please contact us at contact (at) eslint (dot) org and we'll send you the instructions. Thanks!

**Pull request merged**
